File: trunk/tests/t-encrypt.c
/* t-encrypt.c  - regression test
 *	Copyright (C) 2000 Werner Koch (dd9jn)
 *
 * This file is part of GPGME.
 *
 * GPGME is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * GPGME is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

#include "../gpgme/gpgme.h"

#define fail_if_err(a) do { if(a) {                                       \
                               fprintf (stderr, "%s:%d: GpgmeError %s\n", \
                                __FILE__, __LINE__, gpgme_strerror(a));   \
                                exit (1); }                               \
                             } while(0)


int 
main (int argc, char **argv )
{
    GpgmeCtx ctx;
    GpgmeError err;
    GpgmeData in, out;
    GpgmeRecipientSet rset;

    err = gpgme_new_context (&ctx);
    fail_if_err (err);

    err = gpgme_new_data ( &in, "Hallo Leute", 11, 0 );
    fail_if_err (err);

    err = gpgme_new_data ( &out, NULL, 0,0 );
    fail_if_err (err);

    err = gpgme_new_recipient_set (&rset);
    fail_if_err (err);
    err = gpgme_add_recipient (rset, "Bob");
    fail_if_err (err);


    err = gpgme_encrypt (ctx, rset, in, out );
    fail_if_err (err);

   
    gpgme_release_recipient_set (rset);
    gpgme_release_data (in);
    gpgme_release_data (out);
    gpgme_release_context (ctx);
    return 0;
}

File: lang/python/tests/t-encrypt.py
#!/usr/bin/env python

# Copyright (C) 2016 g10 Code GmbH
#
# This file is part of GPGME.
#
# GPGME is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# GPGME is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General
# Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this program; if not, see <https://www.gnu.org/licenses/>.

from __future__ import absolute_import, print_function, unicode_literals

import gpg
import support

del absolute_import, print_function, unicode_literals

c = gpg.Context()
c.set_armor(True)

source = gpg.Data("Hallo Leute\n")
sink = gpg.Data()

keys = []
keys.append(c.get_key("A0FF4590BB6122EDEF6E3C542D727CC768697734", False))
keys.append(c.get_key("D695676BDCEDCC2CDD6152BCFE180B1DA9E3B0B2", False))

c.op_encrypt(keys, gpg.constants.ENCRYPT_ALWAYS_TRUST, source, sink)
result = c.op_encrypt_result()
assert not result.invalid_recipients, \
  "Invalid recipients: {}".format(", ".join(r.fpr for r in result.recipients))
support.print_data(sink)

# Idiomatic interface.
with gpg.Context(armor=True) as c:
    ciphertext, _, _ = c.encrypt(
        "Hallo Leute\n".encode(),
        recipients=keys,
        sign=False,
        always_trust=True)
    assert len(ciphertext) > 0
    assert ciphertext.find(b'BEGIN PGP MESSAGE') > 0, 'Marker not found'

    c.encrypt(
        "Hallo Leute\n".encode(),
        recipients=[c.get_key(support.encrypt_only, False)],
        sign=False,
        always_trust=True)

    try:
        c.encrypt(
            "Hallo Leute\n".encode(),
            recipients=[c.get_key(support.sign_only, False)],
            sign=False,
            always_trust=True)
    except gpg.errors.InvalidRecipients as e:
        assert len(e.recipients) == 1
        assert support.sign_only.endswith(e.recipients[0].fpr)
    else:
        assert False, "Expected an InvalidRecipients error, got none"

    try:
        # People might be tempted to provide strings.
        # We should raise something useful.
        ciphertext, _, _ = c.encrypt(
            "Hallo Leute\n", recipients=keys, sign=False, always_trust=True)
    except TypeError as e:
        # This test is a bit fragile, because the message
        # may very well change. So if the behaviour will change
        # this test can easily be deleted.
        assert "encode" in str(e)
